### PR TITLE
feat(ai): add Kiro provider with OAuth authentication support

### PR DESCRIPTION
**Pull request by mikeyobrien**
## Add Kiro Provider with OAuth Authentication Support

This PR adds support for Amazon's Kiro AI assistant (Amazon Q Developer) as a new provider in the `@pi/ai` package.

### What is Kiro?

Kiro is Amazon's internal AI assistant service that provides access to Claude models (Haiku, Sonnet, Opus) and other models through AWS infrastructure. It uses AWS Builder ID authentication via OAuth device code flow.

### Changes

**Core Implementation:**
- New Kiro provider with streaming support (`packages/ai/src/providers/kiro.ts`)
- OAuth 2.0 device code flow with AWS Builder ID (`packages/ai/src/utils/oauth/kiro.ts`)
- Custom thinking tag parser for extended reasoning mode (`packages/ai/src/providers/kiro-thinking-parser.ts`)
- Model generation script integration for automatic model discovery

**Supported Models:**
- Claude Haiku 4.5
- Claude Sonnet 4, 4.5, 4.5-1M
- Claude Opus 4.5, 4.6
- Qwen3 Coder 480B

**Features:**
- Streaming responses with proper event parsing
- OAuth authentication with automatic token refresh
- Extended thinking mode support with configurable token budgets
- Tool calling with proper message transformation
- Image support in user messages
- History truncation and sanitization for large contexts
- Comprehensive error handling with automatic retries

**Testing:**
- Added Kiro to all cross-provider test suites (10 test files)
- Coverage includes: streaming, abort, empty responses, context overflow, tool calls, images, unicode handling, token counting

**Documentation:**
- Updated `packages/ai/README.md` with Kiro setup instructions
- Added provider documentation in `packages/coding-agent/docs/providers.md`
- Updated coding-agent README with Kiro configuration

### Authentication

Kiro uses AWS Builder ID via OAuth device code flow (same as kiro-cli). Users authenticate by:
1. Running `/login kiro` in pi
2. Opening the provided URL and entering the device code
3. Signing in with Google, GitHub, or AWS credentials

Credentials are automatically refreshed and stored securely.

### Files Changed
- 27 files changed, 1,886 insertions(+), 15 deletions(-)
- New files: `kiro.ts` (987 lines), `kiro-thinking-parser.ts` (193 lines), `oauth/kiro.ts` (210 lines)
- Updated: model generation, type definitions, test suites, documentation

### Testing
All existing tests pass with Kiro integration. The provider handles edge cases including orphaned tool calls, consecutive tool results, history sanitization, and context overflow with automatic retry logic.

### CODE REVIEW

**Comment by badlogic** [comment]
please move kiro-thinking-parser.ts inti kiro.ts, lgtm otherwise.

File: packages/ai/src/utils/oauth/kiro.ts [hunk 1]
```diff
@@ -220,45 +220,100 @@ function getKiroCliToken(): string | undefined {
...
-					const bufferMs = 2 * 60 * 1000; // 2 minutes
-					if (now >= expiresAt - bufferMs) {
-						// Token is expired or about to expire
-						return undefined;
-					}
-				}
-				return data.access_token;
-			}
+
+		// Search for both token and device registration
```
Comment (mikeyobrien) [inline packages/ai/src/utils/oauth/kiro.ts:224]: Using this mess to avoid adding a sqlite dep to pi.

**Comment by mikeyobrien** [comment]
> please move kiro-thinking-parser.ts inti kiro.ts, lgtm otherwise.

@badlogic Done

**Comment by badlogic** [comment]
Hey @mikeyobrien, thanks for putting this together — getting Kiro working as a provider is a solid idea and clearly a lot of work went into this. That said, we can't merge this in its current form. Here's the full list of issues we found:

**Code quality:**
- Dynamic import (`await import("../utils/oauth/kiro.js")`) inside `streamKiro`. We don't allow inline/dynamic imports anywhere in the codebase — all imports must be top-level.
- Module-level mutable state (`let credentials`) in `kiro.ts`. Other providers don't use this pattern and it causes problems with tests and multi-session scenarios.
- Token usage is estimated via `Math.ceil(totalContent.length / 4)` instead of coming from the API. This produces inaccurate costs and misleading usage stats.
- The context window for usage estimation is hardcoded to 200k, but several models have different windows (1M for sonnet-4-5-1m, 131k for qwen3-coder). Usage will be wrong for those.
- `resolveKiroModel()` silently falls back to Sonnet 4.5 for unknown model IDs (including `auto`) instead of erroring.

**SQLite credential extraction:**
- `getKiroCliCredentials()` loads the entire kiro-cli SQLite database (potentially 600MB+) into memory with `readFileSync`, then uses regex on raw binary data to find JSON. This is fragile (the regex `/\{[^\}]+\}/` can't match nested objects), expensive, and runs on every `hasAuth()`/`getApiKey()` call. The commit message mentions switching to sqlite3 CLI but the code still uses `readFileSync`.

**Changelog:**
- The entry was added to the released `[0.52.7]` section instead of `[Unreleased]`. Released sections are immutable.
- Missing PR link and author attribution (required format for external contributions).
- No `packages/coding-agent/CHANGELOG.md` entry for the `auth-storage.ts` changes.

**Unrelated changes:**
- Commits removing "ephemeral agent", "ralph files", and "CODE_TASK files" — unclear what these are and they shouldn't be in this PR.
- Removed opencode models (`glm-4.7-free`, `qwen3-coder`) and openrouter models (`ai21/jamba-mini-1.7`, `stepfun-ai/step3`) as a side effect of regenerating models. Unrelated to Kiro.
- Cosmetic diff noise in `env-api-keys.ts` (removed comment and blank line).

**Testing:**
- No unit tests for `ThinkingTagParser`, `parseKiroEvents`, `buildHistory`, `truncateHistory`, `sanitizeHistory`, or the SQLite credential extraction. These are complex functions with non-trivial edge cases.

**Other:**
- Missing `loginKiro` export from `oauth/index.ts` (all other providers export their login function).
- `OAuthProviderInterface` was extended with `getCliToken()` and `getCliCredentials()` — this is a type-level change that affects all providers, not just Kiro.

---

Rather than trying to fix all of this, we'd suggest turning Kiro into a **pi extension** instead. Extensions can register custom providers without modifying core packages, and it would let you iterate independently. See [docs/extensions.md](https://github.com/badlogic/pi-mono/blob/main/packages/coding-agent/docs/extensions.md) and the [examples](https://github.com/badlogic/pi-mono/tree/main/packages/coding-agent/examples/extensions) for how to set that up.

Thanks again for the effort, and happy to help if you go the extension route.

**Comment by ethaneng** [comment]
@mikeyobrien if you do go down the extension route to integrate Kiro, I'd love to see it if you could mention it here. Thx

**Comment by cosminilie** [comment]
https://github.com/mikeyobrien/pi-provider-kiro is probably the follow up work
